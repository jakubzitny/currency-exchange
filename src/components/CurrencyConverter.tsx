import React, { useEffect, useState } from 'react'
import { List } from 'immutable'
import useSWR from 'swr'

import Wallet from '../entities/wallet'
import { Currency } from '../entities/currency'

import Arrow from './Arrow'
import WalletsProvider from './WalletsProvider'
import WalletsView from './WalletsView'

const API_URL = 'https://api.exchangeratesapi.io/latest'
const API_REFRESH_INTERVAL = 10000

const fetcher = async (key: string) => {
  const response = await fetch(`${API_URL}${key}`)
  return await response.json()
}

const CurrencyConverter: React.FC<{}> = () => {
  const [
    { rate, sourceWalletCurrency, targetWalletCurrency },
    setState,
  ] = useState({
    rate: 1,
    sourceWalletCurrency: Currency.GBP,
    targetWalletCurrency: Currency.USD,
  })
  const { data, error } = useSWR(`?base=${sourceWalletCurrency}`, fetcher, {
    refreshInterval: API_REFRESH_INTERVAL,
  })

  useEffect(() => {
    const rate = data && data.rates && data.rates[targetWalletCurrency]
    if (!rate) {
      return
    }

    setState((prevState) => ({
      ...prevState,
      rate,
    }))
  }, [data])

  const handleSourceWalletCurrencyChange = (currency: Currency) => {
    setState((prevState) => {
      const rate =
        sourceWalletCurrency === currency ? prevState.rate : 1 / prevState.rate
      return {
        ...prevState,
        rate,
        sourceWalletCurrency: currency,
      }
    })
  }

  const handleTargetWalletCurrencyChange = (currency: Currency) => {
    setState((prevState) => {
      const rate =
        (data && data.rates && data.rates[currency]) || prevState.rate
      return {
        ...prevState,
        rate,
        targetWalletCurrency: currency,
      }
    })
  }

  return (
    <WalletsProvider>
      {({ wallets }: { wallets: List<Wallet> }) => (
        <WalletsView
          rate={rate}
          wallets={wallets}
          sourceWalletCurrency={sourceWalletCurrency}
          targetWalletCurrency={targetWalletCurrency}
          onSourceWalletChange={handleSourceWalletCurrencyChange}
          onTargetWalletChange={handleTargetWalletCurrencyChange}
        />
      )}
    </WalletsProvider>
  )
}

export default CurrencyConverter
