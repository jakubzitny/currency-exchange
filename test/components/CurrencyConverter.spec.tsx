import chai, { expect } from 'chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'

import CurrencyConverter from '../../src/components/CurrencyConverter'

chai.use(chaiJestSnapshot)

describe('CurrencyConverter', () => {
  it('should render CurrencyConverter', () => {
    const arrow = <CurrencyConverter />
    const wrapper = shallow(arrow)

    expect(wrapper).to.matchSnapshot()
  })

  it('should fully render CurrencyConverter', (callback) => {
    const arrow = <CurrencyConverter />
    act(() => {
      const wrapper = mount(arrow)
      expect(wrapper).to.matchSnapshot()
      callback()
    })
  })
})
