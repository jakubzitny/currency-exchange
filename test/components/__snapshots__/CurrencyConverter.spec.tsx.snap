// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter should fully render CurrencyConverter 1`] = `ReactWrapper {}`;

exports[`CurrencyConverter should render CurrencyConverter 1`] = `ShallowWrapper {}`;
